---
title: "Mastitis scRNAseq update"
date: '2022-09-15'
output:
  html_document:
    toc: true
    toc_float: true
    toc_depth: 2
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(kableExtra)

```

# Overview  


# Extract Raw data  

## Code  

- scripts/01_tar.slurm  



```{bash, eval=FALSE, code = readLines("scripts/01_tar.slurm")}



```

# Prepare reference genome  

## Code  
  
- scripts/02_cellranger_reference_prep_script.slurm  

```{bash, code=readLines("scripts/02_cellranger_reference_prep_script.slurm"), eval=FALSE}


```

# FastQC  

## Code  
  
- scripts/03_fastQC.slurm  
  
```{bash, code=readLines("scripts/03_fastQC.slurm"), eval=FALSE} 

```

# Cellranger  


## Code {.tabset}  
  
### Cellranger Slurm job array  

- scripts/04_cellranger_map_individual.slurm    
  
```{bash, eval=FALSE, code=readLines("scripts/04_cellranger_map_individual.slurm")}


```

### Cellranger QC metrics  

- This script checks some summary statistics from the Cellranger runs  
- scripts/05_check_mapping.R  
  
```{r, eval=FALSE, code=readLines('scripts/05_check_mapping.R')}

```
  
## Cellranger Summary  

!['cells per replicate'](outputs/Estimated_cells.jpeg)
!['reads per cell'](outputs/Estimated_reads_per_cell.jpeg)  




```{r}
summarydat <- 
  read_tsv('outputs/first_cellranger_results.tsv') %>% 
  select(-`Number of Reads`) %>% 
  mutate(ID=paste0(individual, tissue)) %>% 
  select(ID, everything(),-individual, -tissue,)
  

summarydat %>%  kable() %>% kable_styling(bootstrap_options = 'striped', font_size = 10)
```

  
## SoupX with markers {.tabset}  
  
### 1312blood  
  
#### Pre-SoupX  
  
![''](outputs/figures/1312blood_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1312blood_post_soupX.jpeg)    
  
  
### 1312milk  
  
#### Pre-SoupX  
  
![''](outputs/figures/1312milk_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1312milk_post_soupX.jpeg)    
  
### 1630blood  
  
#### Pre-SoupX  
  
![''](outputs/figures/1630blood_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1630blood_post_soupX.jpeg)  

### 1630milk  
  
#### Pre-SoupX  
  
![''](outputs/figures/1630milk_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1630milk_post_soupX.jpeg)  
  
### 1634blood  
  
#### Pre-SoupX  
  
![''](outputs/figures/1634blood_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1634blood_post_soupX.jpeg)  
  

  
### 1634milk  
  
#### Pre-SoupX  
  
![''](outputs/figures/1634milk_pre_soupX.jpeg)  
  
#### Post-SoupX  
  
![''](outputs/figures/1634milk_post_soupX.jpeg)  
  



